gce/survival-training/README.md

```diff
@@ -0,0 +1,93 @@
+
+
+### Trainer CLI
+
+Our strategy will have the GCE instance automatically kick off your training job on startup. As part of this kick off, it will have to provide the job with the appropriate parameters -- how many steps to train for, how often to checkpoint the model, which data to train the model with, what hyperparameters should be used in training the model, etc.
+
+It is helpful, if you intend to do this more than once, to assert a common semantical system across all your trainers. We provide a [boilerplate CLI](./dummy/train.py) in this repo. This interface is very similar to that provided by [TensorFlow estimators](https://www.tensorflow.org/programmers_guide/estimators), but can be used even with other frameworks. In the sections that follow, we will provide you with both TensorFlow and with scikit-learn examples that demonstrate its use.
+
+
+### Cloud Source Repositories
+
+We need a mechanism by which we can get our trainer code into the GCE instance. There are several solutions available to us:
+
+1. We could simply store the trainer application on the custom image from which we create the GCE instance.
+
+1. We could store the code in a [GCS](https://cloud.google.com/storage/) bucket and have the startup script download it to the instance before doing anything else.
+
+1. We could store the code in a [Cloud Source Repository](https://cloud.google.com/source-repositories/) and have the startup script clone it to the instance before doing anything else.
+
+In this guide, in the interests of flexibility, we will go with the Cloud Source Repository option. The reason this is attractive is that it lets you apply `git` semantics to your training jobs -- for example, you can [tag](https://git-scm.com/book/en/v2/Git-Basics-Tagging) different model architectures and specify them through the instance metadata. Or you could just use the trainer on some special branch of the repository which, again, would be specified at deployment time using instance metadata. This setup is much more powerful than the others.
+
+
+### Startup script
+
+GCE instances have an in-built mechanism that triggers a script on instance startup (for the appropriate definition of startup, but this need not concern us for now). This is the mechanism we will use to kick off a training job or to pick it up where it left off before a shutdown. We will make use of it through [startup scripts](https://cloud.google.com/compute/docs/startupscript#troubleshooting).
+
+It is alright if you are not familiar with these things. What is more important at the moment is to know that they exist and have a general sense for how they fit into the training process.
+
+
+### Instance metadata
```

Comment:
> You may want to deploy different parametrizations of a training job to multiple GCE instances at different times. For this purpose, we will make use of the metadata server available to every GCE instance.

Compute Engine instances have a [metadata server](https://cloud.google.com/compute/docs/storing-retrieving-metadata) that we use to store different groups of parameter settings, or _parameterizations_, for training jobs. This gives you the flexibility to deploy training jobs with different parameterizations to multiple Compute Engine instances at different times.

> Instance metadata is where we will store things like model hyperparameters and paths to training and evaluation data. 

Instance metadata stores training information such as:

- model hyperparameters
- locations of training data
- locations of evaluation data

> We will store metadata on our instances before we start our training jobs, and our startup scripts will make use of this metadata on instance startup.

We will store these types of training information as instance metadata before we start our training jobs. Once the Compute Engine instance starts up, our startup scripts make use of this metadata immediately.

> If you are using the example startup script provided in this guide, you will only have to make minimal changes to do so. The important thing again is to understand where instance metadata fits into our picture and that GCE guarantees that it will be available to our startup script when it runs.

Adapting our example startup script to your training requires minimal changes. Instance metadata and startup scripts work together to start the training process in your Compute Engine instance.

Reply:
Please edit that last part of my edit if I'm missing anything important about the relationship between instance metadata and startup scripts. 

Follow-up question: what parts of the startup script would the user have to change? Do you mean that they just have to change instance metadata and a few things in the startup script? Could you list a couple of quick examples?

Reply:
Thanks for the review @thedriftofwords , will make the changes you recommended in the morning.

To answer your question about required modifications - in this case, I only modified the metadata retrieval and the trainer arguments.

If someone wanted to download the python module from a GCS bucket, they would have to replace the `gcloud source repos clone` bit with `gsutil -m cp -r <GCS BUCKET PATH> .`.

If someone wanted to do a pip install, they would have to add that line into the script.

Hopefully, as we add a few more examples here, this should become clearer.

For now, the diff between startup.sh and tf-estimator-startup.sh is the only real illustration we have. That and imagination.

Reply:
@dizcology : Made the changes you suggested, please have a look and let me know if this is acceptable.